torchao/prototype/uintx/Uintx.py

```diff
@@ -0,0 +1,207 @@
+import functools
+import math
+from collections import defaultdict
+from typing import Any, Callable, Dict, Optional, Tuple, Union, List
+from dataclasses import dataclass
+import torch
+from torch._dynamo.comptime import comptime
+
+from torch.utils._python_dispatch import return_and_correct_aliasing
+from .bitpacking import pack, unpack, numbits
+from torchao.dtypes.utils import (
+    LayoutType,
+    _implements,
+    _register_layout_cls,
+    _dispatch__torch_function__,
+    _dispatch__torch_dispatch__,
+)
+from torchao.dtypes.affine_quantized_tensor import PlainAQTLayout, register_layout_cls
+
+
+aten = torch.ops.aten
+
+class UintxTensor(torch.Tensor):
+    """
+    Splits int data into packed shards based on bit size
+    fields:
+      int4_shard (torch.Tensor): 4 bit packed shard
+      int2_shard (torch.Tensor): 2 bit packed shard
+      int1_shard (torch.Tensor): 1 bit packed shard
+      bit_size (int): element size in bits
+      pack_dim: (int) dimension to pack along
+    """
+    bits_to_shard = {
+        1: ["int1_shard"],
+        2: ["int2_shard"],
+        3: ["int2_shard", "int1_shard"],
+        4: ["int4_shard"],
+        5: ["int4_shard", "int1_shard"],
+        6: ["int4_shard", "int2_shard"],
+        7: ["int4_shard", "int2_shard", "int1_shard"],
+    }
+    def __new__(
+        cls,
+        shards: List[torch.Tensor],
+        packed_shape: List[int],
+        bit_size: int,
+        pack_dim: int = -1,
+    ):
+        kwargs = {"device": shards[0].device}
+        kwargs["device"] = shards[0].device
+        kwargs["layout"] = shards[0].layout
+        kwargs["requires_grad"] = False
+        kwargs["dtype"] = torch.uint8
+        return torch.Tensor._make_wrapper_subclass(cls, packed_shape, **kwargs) 
+
+    def __init__(
+        self,
+        shards: List[torch.Tensor],
+        packed_shape: List[int],
+        bit_size: int,
+        pack_dim: int = -1,
+    ):
+        for i, attrib in enumerate(self.bits_to_shard[bit_size]):
+            setattr(self, attrib, shards[i])
+            
+        self.packed_shape = packed_shape
+        self.bit_size = bit_size    
```

Comment:
also I feel it would be helpful to add an assert for accepted `bit_size`, also mayby rename this to `bit_width`? feels more commen

Reply:
What would the assert be checking?

Reply:
it should one of 1-7 right?